.changeset/wise-llamas-drop.md

```diff
@@ -0,0 +1,137 @@
+---
+'@graphql-yoga/plugin-jwt': major
+---
+
+Re-write for the JWT plugin. This plugin can be configured now with multiple providers, lookup locations, token verification, and more. 
+
+The version has better version coverage, and it provides an improved API for configuring provider and custom behaviors.
+
+## Breaking Change: New Plugin Configuration
+
+### Signing key providers
+
+❌ The `signingKey` option has be removed.
+❌ The `jwksUri` + `jwksOpts` options has been removed.
+✅ Multiple signing key providers and support for fallbacks (`singingKeyProviders[]`).
+✅ Improved API for defining signing key configuration.
+✅ Better defaults for caching and rate-limiting for remote JWKS providers. 
+
+#### Before
+
+```ts
+useJWT({
+  signingKey: "...",
+  // or
+  jwksUri: "http://example.com/..."
+  jwksOpts: {
+    // ... 
+  }
+})
+```
+
+#### After
+
+```ts
+useJWT({
+  // Pass one or more providers
+  singingKeyProviders: [
+    {
+      type: "remote",
+      jwksClientOptions: {
+        // ... pass options to https://github.com/auth0/node-jwks-rsa 
+      }
+    },
+    // This one also acts as a fallback in case of a fetching issue with the 1st provider
+    {
+      type: "inline",
+      signingKey: "..." 
+    }
+  ]
+})
+```
+
+### Improved Token Lookup 
+
+❌ Removed `getToken` option from the root config. 
+✅ Added support for autmatically extracting the JWT token from cookie or header.
+✅ Easier setup for extracting from multiple locations
```

Comment:
To be honest, I find this way of declaratively configure the token extraction a bit easier to understand. And it doesn't require string manipulation :-)

Do you have some use cases in mind where the string based notation could allow more things than here ?

Reply:
It is a discussion for Mesh. I am just saying if the idea is simplifying Mesh usage, this is not the right way.
I think functions are better for the libraries like Yoga, because users are able use the default logic from the lib itself.

```ts
getToken: extractTokenFromHeader({ prefix: 'Bearer', name: 'authorization' })
```

In this case, `extractTokenFromHeader` is a function you can re-use within your custom logic for example. That's why I prefer functions over declarative API.

Reply:
Fixed